Implementing sorting algorithms in JavaScript (#1239)
section-engineering-education/engineering-education: articles/sorting-algorithms-in-js/index.md

Sorting can be referred to as an operation performed to arrange records in some particular order. The arrangement is performed based on the value of each record present. The order applied can either be ascending or descending. Sorting algorithms are instructions given to a computer to arrange elements in a particular order.

### Prerequisites

- You should have some basic knowledge of JavaScript.

- You should have [Node.js](#https://nodejs.org/en/) installed on your computer.

### Overview

- [Categories of sorting](#categories-of-sorting)

- [Efficiency of sorting algorithms](#efficiency-of-sorting-algorithms)

- [Sorting algorithms](#sorting-algorithms)

### Categories of sorting

Sorting can be categorized into two:

- **Internal sorting**: It is where only the main memory is used in adjusting the data.

> - **External sorting**: It is where you apply external storage devices such as hard drives, flash disks, etc so that the data can be adjusted.

Author: fixed.

Based on the above factors, an algorithm has three performance cases:

- **Worst case time complexity**: It is a function defined as a result of a maximum number of steps taken on any instance of size n. It is usually expressed in [Big O notation](https://www.geeksforgeeks.org/analysis-of-algorithms-set-3asymptotic-notations/).

- **Average case time complexity**: It is a function defined as a result of the average number of steps taken on any instance of size n. It is usually expressed in [Big theta notation](https://www.geeksforgeeks.org/analysis-of-algorithms-set-3asymptotic-notations/).

- **Best case time complexity**: It is a function defined as a result of the minimum number of steps taken on any instance of size n. It is usually expressed in [Big omega notation](https://www.geeksforgeeks.org/analysis-of-algorithms-set-3asymptotic-notations/).

- **Space complexity**: It is a function defined as a result of additional memory space needed to carry out the algorithm. It is usually expressed in [Big O notation](https://www.geeksforgeeks.org/analysis-of-algorithms-set-3asymptotic-notations/).

### Strategies applied during sorting

> - **Recursion**: The algorithm accomplishes its task by repeating a particular process in a particular order.

Author: fixed.

> - **Divide and conquer**: The algorithm accomplished its task by dividing the data into smaller portions so that to handle it.

Author: fixed.

### Sorting algorithms

For each sorting algorithm discussed below, there is: a step-by-step explanation of how the algorithm works, pictorial representation, and implementation of the algorithm using JavaScript. 

### Bubble sort.

Bubble sort follows the recursion technique.

**Step-by-step guide**:

- Start by comparing two elements in a list.

- Swap them if required.

- Continue till the end of the list. At this point, you have made a series of inner passes and completed an outer pass.

> - Repeat the process until the entire list is sorted.

Author: fixed.

**Pictorial representation**:

![bubble_sort](/engineering-education/sorting-algorithms-in-js/bubble-sort.png)

**JavaScript implementation**.

```JavaScript
function bubbleSort(arr){

    //Outer pass
    for(let i = 0; i < arr.length; i++){

        //Inner pass
        for(let j = 0; j < arr.length - i - 1; j++){

            //Value comparison using ascending order
            
            if(arr[j + 1] < arr[j]){

                //Swapping
                [arr[j + 1],arr[j]] = [arr[j],arr[j + 1]]
            }
        }
    };
    return arr;
};

console.log(bubbleSort([5,3,8,4,6]));
```

Output

```bash
[3,4,5,6,8]
```

Bubble sort has the following performance cases:

- Worst-case time complexity of [`Big O`](https://www.geeksforgeeks.org/difference-between-big-oh-big-omega-and-big-theta/) n ^2. Requires n steps to sort a list of n elements.

- Average-case time complexity of [`Big theta`](https://www.geeksforgeeks.org/difference-between-big-oh-big-omega-and-big-theta/) n ^ 2. Requires n steps to sort a list of n elements.

> - Best-case time complexity of [`Big omega`](https://www.geeksforgeeks.org/difference-between-big-oh-big-omega-and-big-theta/): O(n). Requires n steps to sort a sorted list of n elements.

Reviewer: You don't have to provide the same links everywhere. Just provide it at the intro where you are talking about the notations.

Author: fixed.

- Worst-case time complexity: [`Big O`](https://www.geeksforgeeks.org/difference-between-big-oh-big-omega-and-big-theta/) n * log n. It divides a list into two halves and takes linear time to merge the two halves.

- Average-case time complexity: [`Big theta`](https://www.geeksforgeeks.org/difference-between-big-oh-big-omega-and-big-theta/) n * log n. It divides a list into two halves and takes linear time to merge the two halves.

- Best-case time complexity: [`Big omega`](https://www.geeksforgeeks.org/difference-between-big-oh-big-omega-and-big-theta/) n * log n. It divides a list into two halves and takes linear time to merge the two halves. 

- Space complexity: [`Big-O`](https://www.geeksforgeeks.org/difference-between-big-oh-big-omega-and-big-theta/) n. It requires additional space based on the elements of the list. It is not preferred for large lists.

### Quick sort

Quicksort applies the divide and conquer technique. It works by having a pivot element such that the elements to the left of it are less than it and those to the right are greater than it. The pivot element can be any element in the list.

**Step-by-step guide**:

> - Identify the pivot element.

Reviewer: `Identify` is misleading, you can say select a pivot element.

Author: fixed.

- Increment the right and the left pointer. At this point, a further increment shall cause the left pointer to be greater than the right pointer which shall break the loop returning the left pointer.

- Recursively do the above until the entire list is sorted.

Quicksort has the following performance cases:

- Worst-case time complexity : [`Big O`](https://www.geeksforgeeks.org/difference-between-big-oh-big-omega-and-big-theta/) n^2. In the case of unbalanced subarrays, it will take more time to finish.

- Average-case time complexity: [`Big theta`](https://www.geeksforgeeks.org/difference-between-big-oh-big-omega-and-big-theta/) n * log n. It divides an array into two and shall take linear time to merge sorted portions.

- Best-case time complexity: [`Big omega`](https://www.geeksforgeeks.org/difference-between-big-oh-big-omega-and-big-theta/) n * log n. Given  a sorted array, it shall divide the array into two and shall take linear time to merge sorted portions.

- Space complexity: [`Big O`](https://www.geeksforgeeks.org/difference-between-big-oh-big-omega-and-big-theta/) n * log n. Requires small additional space to hold the subarrays.

### Key takeaways

> - JavaScript by default uses insertion sort for the `sort()`. This means that it is not appropriate when sorting large lists. Therefore when dealing with large lists you should consider other sorting algorithms such as quicksort or merge sort.

Reviewer: ```suggestion
- JavaScript by default uses insertion sort for the `sort()` method. This means that it is not appropriate when sorting large lists. Therefore when dealing with large lists you should consider other sorting algorithms such as quicksort or merge sort.
```

Author: fixed.

> - Divide and conquer based sorting algorithm are faster than Recursion based sorting algorithm.

Reviewer: ```suggestion
- Generally, a Divide and conquer based sorting algorithm is faster than Recursion based sorting algorithm.
```

Author: fixed.

### Conclusion

> With an understanding of sorting algorithms, software developers can be able to realize the appropriate algorithm to use depending on the data set, the time required and the space available. This is without limitation of the default sorting algorithm provided by the language.

Reviewer: ```suggestion
With an understanding of sorting algorithms, software developers can be able to realize the appropriate sorting algorithm to use depending on the data set, the time required and the space available.
```

Author: fixed.

Sorting can be referred to as an operation performed to arrange records in some particular order. The arrangement is performed based on the value of each record present. The order applied can either be ascending or descending. Sorting algorithms are instructions given to a computer to arrange elements in a particular order.

### Prerequisites

- You should have some basic knowledge of JavaScript.

- You should have [Node.js](#https://nodejs.org/en/) installed on your computer.

### Overview

- [Categories of sorting algorithms](#categories-of-sorting-algorithms)

- [Efficiency of sorting algorithms](#efficiency-of-sorting-algorithms)

> - [Sorting algorithms](#sorting-algorithms)

Reviewer: In the Overview section, please add the list of sorting algorithms you are discussing in the article.

Author: fixed.
